prog/heatmap.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package prog
+
+import (
+	"bytes"
+	"fmt"
+	"math/rand"
+)
+
+// Our heatmaps are a flexible mechanism to assign a probability distribution to
+// some collection of bytes. Usage:
+//  1. Choose a heatmap and initialize it: `hm := MakeXYZHeatmap(data)`.
+//     Different heatmaps implement different probability distributions
+//     (for now there is only one).
+//  2. Select random indices according to the probability distribution:
+//     `idx := hm.ChooseLocation(r)`.
+type Heatmap interface {
+	ChooseLocation(r *rand.Rand) int
+}
+
+// Generic heatmaps model a probability distribution based on sparse data,
+// prioritising selection of regions which are not a single repeated byte. It
+// views data as a series of chunks of length `granularity`, ignoring chunks
+// which are a single repeated byte. Indices are chosen uniformly amongst the
+// remaining "interesting" segments.
+func MakeGenericHeatmap(data []byte) Heatmap {
+	var hm GenericHeatmap
+	hm.rawLength = len(data)
+	hm.length, hm.segments = calculateLengthAndSegments(data, granularity)
+	return &hm
+}
+
+func (hm *GenericHeatmap) ChooseLocation(r *rand.Rand) int {
+	if hm.length == 0 {
```

Comment:
An alternative may be to always add a segment covering all data, if we got no segments. But not sure which is better.

Reply:
I think this suggestion makes more sense, so I will update the implementation.